pkg/pillar/cmd/zedagent/handlecertconfig.go

```diff
@@ -127,70 +142,181 @@ func unpublishControllerCert(ctx *getconfigContext, key string) {
+	if err != nil {
+		log.Errorf("getCertsFromController: verify err %v", err)
+		return false
+	}
+	err = fileutils.WriteRename(types.ServerSigningCertFileName, certBytes)
+	if err != nil {
+		log.Errorf("getCertsFromController: file save err %v", err)
+		return false
+	}
+
+	log.Infof("getCertsFromController: success")
+	return true
+}
+
+// eve node certificate post task, on change trigger
+func eveNodeCertsTask(ctx *zedagentContext, triggerEveNodeCerts chan struct{}) {
+	log.Infoln("starting eve node certificates publish task")
+
+	publishEveNodeCertsToController(ctx)
+
+	stillRunning := time.NewTicker(25 * time.Second)
+	agentlog.StillRunning(agentName+"attest", warningTime, errorTime)
+
+	for {
+		select {
+		case <-triggerEveNodeCerts:
+			start := time.Now()
+			publishEveNodeCertsToController(ctx)
+			pubsub.CheckMaxTimeTopic(agentName+"attest",
+				"publishEveNodeCertsToController", start,
+				warningTime, errorTime)
+
+		case <-stillRunning.C:
+		}
+		agentlog.StillRunning(agentName+"attest", warningTime, errorTime)
+	}
+}
+
+// prepare the eve node certs list proto message
+func publishEveNodeCertsToController(ctx *zedagentContext) {
 	var attestReq = &attest.ZAttestReq{}
 
+	// not V2API
+	if !zedcloud.UseV2API() {
+		return
+	}
+
+	attestReq = new(attest.ZAttestReq)
+	startPubTime := time.Now()
 	attestReq.ReqType = attest.ZAttestReqType_ATTEST_REQ_CERT
 	// no quotes
 
-	// TBD:XXX, get the ECDH Certs here
+	h := sha256.New()
+	sub := ctx.subEveNodeCert
+	items := sub.GetAll()
+	for _, item := range items {
+		config := item.(types.EveNodeCert)
+		computeConfigElementSha(h, item)
+		certMsg := zcert.ZCert{
+			HashAlgo: convertLocalToApiHashAlgo(config.HashAlgo),
+			Type:     convertLocalToApiCertType(config.CertType),
+			Cert:     config.Cert,
+			CertHash: config.CertID,
+		}
+		attestReq.Certs = append(attestReq.Certs, &certMsg)
+	}
+	ctx.cipherCtx.eveNodeCertHash = h.Sum(nil)
 
-	log.Debugf("publishCerts to ZedCloud, sending %s", attestReq)
-	sendCertsProtobuf(attestReq, iteration)
+	log.Debugf("publishEveNodeCertsToController, sending %s", attestReq)
+	sendEveNodeCertsProtobuf(attestReq, ctx.cipherCtx.iteration)
+	log.Debugf("publishEveNodeCertsToController: after send, total elapse sec %v",
+		time.Since(startPubTime).Seconds())
+	ctx.cipherCtx.iteration++
 }
 
 // Try all (first free, then rest) until it gets through.
 // Each iteration we try a different port for load spreading.
 // For each port we try all its local IP addresses until we get a success.
-func sendCertsProtobuf(attestReq *attest.ZAttestReq, iteration int) {
+func sendEveNodeCertsProtobuf(attestReq *attest.ZAttestReq, iteration int) {
 	data, err := proto.Marshal(attestReq)
 	if err != nil {
 		log.Fatal("SendInfoProtobufStr proto marshaling error: ", err)
 	}
 
-	deviceUUID := zcdevUUID.String()
-	zedcloud.RemoveDeferred("attest:" + deviceUUID)
+	reqID := zcdevUUID.String()
+	zedcloud.RemoveDeferred("attest:" + reqID)
```

Comment:
I would be more maintainable if you defined a local deferKey := "attest:"+reqID and use it here plus in the SetDeferred below.
Non-obvious things would happen if the two ever get out of synch due to some future code changes.

Reply:
sure